### ISSUE TITLE
Support versioned shared libraries on MacOS

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#3241]
* Version of SCons
3.0.1
* Version of Python
Any

* What Platform are you on? (Linux/Windows and which version)
MacOS

Initially reported by MacPorts Ryan Schmidt

Discussions include:
- http://mail-archives.apache.org/mod_mbox/serf-dev/201805.mbox/%3c334427a4-3e1e-6445-2f81-d3dbe8ec30b6@apache.org%3e
- https://pairlist4.pair.net/pipermail/scons-users/2018-May/006918.html
- https://pairlist4.pair.net/pipermail/scons-users/2018-October/007328.html

# Issue 
Currently there is no support for SHLIBVERSION on mac. It was likely dropped around 2.4.1 when the current version of SHLIBVERSION logic was rolled out.

MacPorts Ticket: https://trac.macports.org/ticket/50854

Suggested implementation by @acmorrow

> 
> My recommendation would be to not attempt to automatically provide values for -compatibility_version and -current_version, but to provide variables that allow users to set them. If you did want something that mapped SHLIBVERSION=x.y.z to -current_version=y.z -compat_version=y.0 or something (or the +1 variants of that), I think you could do it with generator functions that would parse SHLIBVERSION.
> 
```
# No flags passed for -current_version or -compat_version
 env.SharedLibrary(..., SHLIBVERSION=1.2.3, ...)
 
# User sets explicitly, links with -Wl,-current_version=7.1 -Wl,-compatibility_version=7.0
 env.SharedLibrary(..., SHLIBVERSION=1.2.3, APPLELINK_CURRENT_VERSION=7.1,  APPLELINK_COMPATIBILITY_VERSION=7.0)
 
# User programmatically describes how SHLIBVERSION maps to values for compat/current.

 def currentVersionFromSoVersion(source, target, env, for_signature):
     return env['SHLIBVERSION'].split(.)[1:]
 
 def compatVersionFromSoVersion(source, target, env, for_signature):
     return env['SHLIBVERSION'].split(.)[1] + ".0"
 
 env['APPLELINK_CURRENT_VERSION'] = currentVersionFromSoVersion
 env['APPLELINK_COMPATIBILITY_VERSION'] = compatVersionFromSoVersion
 
# Links as -Wl,current_version=2.3 -Wl,compat_version=2.0
 env.SharedLibrary(..., SHLIBVERSION=1.2.3, ...)
 
# Something like that? Would currentVersionFromSoVersion be able to see the SHLIBVERSION set in the following:
 
 env.SharedLibrary(
     SHLIBVERSION=1.2.3,
     APPLELINK_CURRENT_VERSION=currentVersionFromSoVersion
 )
```
> 
> You could also add some validation for the rules for current_version and compatibility_version in whatever piece of applelink.py ends up doing the expansion of APPLELINK_CURRENT_VERSION and APPLELINK_COMPATIBILITY_VERSION.

**Comment by bdbaddog** [comment]
Further investigation yields that there doesn't seem to be a definitive pattern for versioned shared library symlinks.
It's not uncommon to match scons' current symlink pattern so no need to add special logic just for macOS.

**Comment by bdbaddog** [comment]
Fixed in 3.0.2
